[BUG] org.opensearch.telemetry.tracing.DefaultTracerTests.testSpanAcrossThreadsMultipleSpans is flaky

**Describe the bug**
The org.opensearch.telemetry.tracing.DefaultTracerTests.testSpanAcrossThreadsMultipleSpans is flaky

```
org.opensearch.telemetry.tracing.DefaultTracerTests.testSpanAcrossThreadsMultipleSpans

java.lang.AssertionError: expected:&lt;org.opensearch.test.telemetry.tracing.MockSpan@19146f98&gt; but was:&lt;null&gt;
	at __randomizedtesting.SeedInfo.seed([D9BCB54305D17127:BF4A5911DBDD9525]:0)
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at org.opensearch.telemetry.tracing.DefaultTracerTests.testSpanAcrossThreadsMultipleSpans(DefaultTracerTests.java:385)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)</pre>org.opensearch.telemetry.tracing.DefaultTracerTests.testSpanAcrossThreadsMultipleSpans
```

**To Reproduce**
```
 ./gradlew ':libs:opensearch-telemetry:test' --tests "org.opensearch.telemetry.tracing.DefaultTracerTests.testSpanAcrossThreadsMultipleSpans" -Dtests.seed=D9BCB54305D17127 
```

**Expected behavior**
Test must always pass

**Plugins**
Standard

**Screenshots**
N/A

**Host/Environment (please complete the following information):**
 - CI
 - 
**Additional context**
https://build.ci.opensearch.org/job/gradle-check/24505/testReport/junit/org.opensearch.telemetry.tracing/DefaultTracerTests/testSpanAcrossThreadsMultipleSpans/

--- a/libs/telemetry/src/test/java/org/opensearch/telemetry/tracing/DefaultTracerTests.java
+++ b/libs/telemetry/src/test/java/org/opensearch/telemetry/tracing/DefaultTracerTests.java
@@ -19,7 +19,9 @@ import org.opensearch.test.telemetry.tracing.MockTracingTelemetry;
 import org.opensearch.threadpool.ThreadPool;
 
 import java.io.IOException;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicReference;
 
 import static org.mockito.ArgumentMatchers.any;
@@ -249,7 +251,7 @@ public class DefaultTracerTests extends OpenSearchTestCase {
      * 4. verify the current_span is still the same on async thread as the 2
      * 5. verify the main thread has current span as null.
      */
-    public void testSpanAcrossThreads() {
+    public void testSpanAcrossThreads() throws ExecutionException, InterruptedException {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
         AtomicReference<Span> currentSpanRefThread1 = new AtomicReference<>();
@@ -265,32 +267,41 @@ public class DefaultTracerTests extends OpenSearchTestCase {
         );
         DefaultTracer defaultTracer = new DefaultTracer(tracingTelemetry, spanTracerStorage);
 
-        executorService.execute(() -> {
+        Future<?> asyncTask = executorService.submit(() -> {
             // create a span
             Span span = defaultTracer.startSpan(new SpanCreationContext("span_name_t_1", Attributes.EMPTY));
             SpanScope spanScope = defaultTracer.withSpanInScope(span);
             spanRef.set(span);
 
-            executorService.execute(() -> {
+            Future<?> asyncTask1 = executorService.submit(() -> {
                 Span spanT2 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
                 SpanScope spanScopeT2 = defaultTracer.withSpanInScope(spanT2);
                 spanT2Ref.set(spanT2);
 
                 currentSpanRefThread2.set(defaultTracer.getCurrentSpan().getSpan());
 
-                spanT2.endSpan();
                 spanScopeT2.close();
+                spanT2.endSpan();
                 currentSpanRefAfterEndThread2.set(getCurrentSpanFromContext(defaultTracer));
             });
+            try {
+                asyncTask1.get();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            } catch (ExecutionException e) {
+                throw new RuntimeException(e);
+            }
             spanScope.close();
+            span.endSpan();
             currentSpanRefThread1.set(getCurrentSpanFromContext(defaultTracer));
         });
+        asyncTask.get();
         assertEquals(spanT2Ref.get(), currentSpanRefThread2.get());
-        assertEquals(spanRef.get(), currentSpanRefAfterEndThread2.get());
+        assertEquals(null, currentSpanRefAfterEndThread2.get());
         assertEquals(null, currentSpanRefThread1.get());
     }
 
-    public void testSpanCloseOnThread2() {
+    public void testSpanCloseOnThread2() throws ExecutionException, InterruptedException {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
         ThreadContextBasedTracerContextStorage spanTracerStorage = new ThreadContextBasedTracerContextStorage(
@@ -302,11 +313,14 @@ public class DefaultTracerTests extends OpenSearchTestCase {
         DefaultTracer defaultTracer = new DefaultTracer(tracingTelemetry, spanTracerStorage);
         final Span span = defaultTracer.startSpan(new SpanCreationContext("span_name_t1", Attributes.EMPTY));
         try (SpanScope spanScope = defaultTracer.withSpanInScope(span)) {
-            executorService.execute(() -> async(new ActionListener<Boolean>() {
+            Future<?> asyncTask = executorService.submit(() -> async(new ActionListener<Boolean>() {
                 @Override
                 public void onResponse(Boolean response) {
-                    span.endSpan();
-                    currentSpanRefThread2.set(defaultTracer.getCurrentSpan());
+                    try (SpanScope s = defaultTracer.withSpanInScope(span)) {
+                        currentSpanRefThread2.set(defaultTracer.getCurrentSpan());
+                    } finally {
+                        span.endSpan();
+                    }
                 }
 
                 @Override
@@ -314,9 +328,10 @@ public class DefaultTracerTests extends OpenSearchTestCase {
 
                 }
             }));
+            asyncTask.get();
             currentSpanRefThread1.set(defaultTracer.getCurrentSpan());
         }
-        assertEquals(null, currentSpanRefThread2.get());
+        assertEquals(span, currentSpanRefThread2.get().getSpan());
         assertEquals(span, currentSpanRefThread1.get().getSpan());
         assertEquals(null, defaultTracer.getCurrentSpan());
     }
@@ -334,7 +349,7 @@ public class DefaultTracerTests extends OpenSearchTestCase {
      * 6. verify the current_span is still the same on async thread as the 2
      * 7. verify the main thread has current span as null.
      */
-    public void testSpanAcrossThreadsMultipleSpans() {
+    public void testSpanAcrossThreadsMultipleSpans() throws ExecutionException, InterruptedException {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
         AtomicReference<Span> currentSpanRefThread1 = new AtomicReference<>();
@@ -351,7 +366,7 @@ public class DefaultTracerTests extends OpenSearchTestCase {
         );
         DefaultTracer defaultTracer = new DefaultTracer(tracingTelemetry, spanTracerStorage);
 
-        executorService.execute(() -> {
+        Future<?> asyncTask = executorService.submit(() -> {
             // create a parent span
             Span parentSpan = defaultTracer.startSpan(new SpanCreationContext("p_span_name", Attributes.EMPTY));
             SpanScope parentSpanScope = defaultTracer.withSpanInScope(parentSpan);
@@ -361,28 +376,37 @@ public class DefaultTracerTests extends OpenSearchTestCase {
             SpanScope spanScope = defaultTracer.withSpanInScope(span);
             spanRef.set(span);
 
-            executorService.execute(() -> {
+            Future<?> asyncTask1 = executorService.submit(() -> {
                 Span spanT2 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
                 SpanScope spanScopeT2 = defaultTracer.withSpanInScope(spanT2);
-
                 Span spanT21 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
-                SpanScope spanScopeT21 = defaultTracer.withSpanInScope(spanT2);
+                SpanScope spanScopeT21 = defaultTracer.withSpanInScope(spanT21);
                 spanT2Ref.set(spanT21);
                 currentSpanRefThread2.set(defaultTracer.getCurrentSpan().getSpan());
-
-                spanT21.endSpan();
                 spanScopeT21.close();
+                spanT21.endSpan();
 
-                spanT2.endSpan();
                 spanScopeT2.close();
+                spanT2.endSpan();
+
                 currentSpanRefAfterEndThread2.set(getCurrentSpanFromContext(defaultTracer));
             });
+            try {
+                asyncTask1.get();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            } catch (ExecutionException e) {
+                throw new RuntimeException(e);
+            }
             spanScope.close();
+            span.endSpan();
             parentSpanScope.close();
+            parentSpan.endSpan();
             currentSpanRefThread1.set(getCurrentSpanFromContext(defaultTracer));
         });
+        asyncTask.get();
         assertEquals(spanT2Ref.get(), currentSpanRefThread2.get());
-        assertEquals(spanRef.get(), currentSpanRefAfterEndThread2.get());
+        assertEquals(null, currentSpanRefAfterEndThread2.get());
         assertEquals(null, currentSpanRefThread1.get());
     }
 

--- a/libs/telemetry/src/test/java/org/opensearch/telemetry/tracing/DefaultTracerTests.java
+++ b/libs/telemetry/src/test/java/org/opensearch/telemetry/tracing/DefaultTracerTests.java
@@ -20,7 +20,6 @@ import org.opensearch.threadpool.ThreadPool;
 
 import java.io.IOException;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 
 import static org.mockito.ArgumentMatchers.any;
@@ -250,7 +249,7 @@ public class DefaultTracerTests extends OpenSearchTestCase {
      * 4. verify the current_span is still the same on async thread as the 2
      * 5. verify the main thread has current span as null.
      */
-    public void testSpanAcrossThreads() throws ExecutionException, InterruptedException {
+    public void testSpanAcrossThreads() {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
 
@@ -279,10 +278,10 @@ public class DefaultTracerTests extends OpenSearchTestCase {
             span.endSpan();
             assertEquals(null, defaultTracer.getCurrentSpan());
         }, executorService);
-        asyncTask.get();
+        asyncTask.join();
     }
 
-    public void testSpanCloseOnThread2() throws ExecutionException, InterruptedException {
+    public void testSpanCloseOnThread2() {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
         ThreadContextBasedTracerContextStorage spanTracerStorage = new ThreadContextBasedTracerContextStorage(
@@ -308,7 +307,7 @@ public class DefaultTracerTests extends OpenSearchTestCase {
                 }
             }), executorService);
             assertEquals(span, defaultTracer.getCurrentSpan().getSpan());
-            asyncTask.get();
+            asyncTask.join();
         }
         assertEquals(null, defaultTracer.getCurrentSpan());
     }
@@ -326,7 +325,7 @@ public class DefaultTracerTests extends OpenSearchTestCase {
      * 6. verify the current_span is still the same on async thread as the 2
      * 7. verify the main thread has current span as null.
      */
-    public void testSpanAcrossThreadsMultipleSpans() throws ExecutionException, InterruptedException {
+    public void testSpanAcrossThreadsMultipleSpans() {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
 
@@ -367,7 +366,7 @@ public class DefaultTracerTests extends OpenSearchTestCase {
             parentSpan.endSpan();
             assertEquals(null, defaultTracer.getCurrentSpan());
         }, executorService);
-        asyncTask.get();
+        asyncTask.join();
     }
 
     public void testClose() throws IOException {

--- a/libs/telemetry/src/test/java/org/opensearch/telemetry/tracing/DefaultTracerTests.java
+++ b/libs/telemetry/src/test/java/org/opensearch/telemetry/tracing/DefaultTracerTests.java
@@ -19,10 +19,9 @@ import org.opensearch.test.telemetry.tracing.MockTracingTelemetry;
 import org.opensearch.threadpool.ThreadPool;
 
 import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Future;
-import java.util.concurrent.atomic.AtomicReference;
 
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
@@ -254,12 +253,6 @@ public class DefaultTracerTests extends OpenSearchTestCase {
     public void testSpanAcrossThreads() throws ExecutionException, InterruptedException {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
-        AtomicReference<Span> currentSpanRefThread1 = new AtomicReference<>();
-        AtomicReference<Span> currentSpanRefThread2 = new AtomicReference<>();
-        AtomicReference<Span> currentSpanRefAfterEndThread2 = new AtomicReference<>();
-
-        AtomicReference<Span> spanRef = new AtomicReference<>();
-        AtomicReference<Span> spanT2Ref = new AtomicReference<>();
 
         ThreadContextBasedTracerContextStorage spanTracerStorage = new ThreadContextBasedTracerContextStorage(
             threadContext,
@@ -267,38 +260,26 @@ public class DefaultTracerTests extends OpenSearchTestCase {
         );
         DefaultTracer defaultTracer = new DefaultTracer(tracingTelemetry, spanTracerStorage);
 
-        Future<?> asyncTask = executorService.submit(() -> {
+        CompletableFuture<?> asyncTask = CompletableFuture.runAsync(() -> {
             // create a span
             Span span = defaultTracer.startSpan(new SpanCreationContext("span_name_t_1", Attributes.EMPTY));
             SpanScope spanScope = defaultTracer.withSpanInScope(span);
-            spanRef.set(span);
 
-            Future<?> asyncTask1 = executorService.submit(() -> {
+            CompletableFuture<?> asyncTask1 = CompletableFuture.runAsync(() -> {
                 Span spanT2 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
                 SpanScope spanScopeT2 = defaultTracer.withSpanInScope(spanT2);
-                spanT2Ref.set(spanT2);
-
-                currentSpanRefThread2.set(defaultTracer.getCurrentSpan().getSpan());
+                assertEquals(spanT2, defaultTracer.getCurrentSpan().getSpan());
 
                 spanScopeT2.close();
                 spanT2.endSpan();
-                currentSpanRefAfterEndThread2.set(getCurrentSpanFromContext(defaultTracer));
-            });
-            try {
-                asyncTask1.get();
-            } catch (InterruptedException e) {
-                throw new RuntimeException(e);
-            } catch (ExecutionException e) {
-                throw new RuntimeException(e);
-            }
+                assertEquals(null, defaultTracer.getCurrentSpan());
+            }, executorService);
+            asyncTask1.join();
             spanScope.close();
             span.endSpan();
-            currentSpanRefThread1.set(getCurrentSpanFromContext(defaultTracer));
-        });
+            assertEquals(null, defaultTracer.getCurrentSpan());
+        }, executorService);
         asyncTask.get();
-        assertEquals(spanT2Ref.get(), currentSpanRefThread2.get());
-        assertEquals(null, currentSpanRefAfterEndThread2.get());
-        assertEquals(null, currentSpanRefThread1.get());
     }
 
     public void testSpanCloseOnThread2() throws ExecutionException, InterruptedException {
@@ -308,16 +289,14 @@ public class DefaultTracerTests extends OpenSearchTestCase {
             threadContext,
             tracingTelemetry
         );
-        AtomicReference<SpanContext> currentSpanRefThread1 = new AtomicReference<>();
-        AtomicReference<SpanContext> currentSpanRefThread2 = new AtomicReference<>();
         DefaultTracer defaultTracer = new DefaultTracer(tracingTelemetry, spanTracerStorage);
         final Span span = defaultTracer.startSpan(new SpanCreationContext("span_name_t1", Attributes.EMPTY));
         try (SpanScope spanScope = defaultTracer.withSpanInScope(span)) {
-            Future<?> asyncTask = executorService.submit(() -> async(new ActionListener<Boolean>() {
+            CompletableFuture<?> asyncTask = CompletableFuture.runAsync(() -> async(new ActionListener<Boolean>() {
                 @Override
                 public void onResponse(Boolean response) {
                     try (SpanScope s = defaultTracer.withSpanInScope(span)) {
-                        currentSpanRefThread2.set(defaultTracer.getCurrentSpan());
+                        assertEquals(span, defaultTracer.getCurrentSpan().getSpan());
                     } finally {
                         span.endSpan();
                     }
@@ -327,12 +306,10 @@ public class DefaultTracerTests extends OpenSearchTestCase {
                 public void onFailure(Exception e) {
 
                 }
-            }));
+            }), executorService);
+            assertEquals(span, defaultTracer.getCurrentSpan().getSpan());
             asyncTask.get();
-            currentSpanRefThread1.set(defaultTracer.getCurrentSpan());
         }
-        assertEquals(span, currentSpanRefThread2.get().getSpan());
-        assertEquals(span, currentSpanRefThread1.get().getSpan());
         assertEquals(null, defaultTracer.getCurrentSpan());
     }
 
@@ -352,13 +329,6 @@ public class DefaultTracerTests extends OpenSearchTestCase {
     public void testSpanAcrossThreadsMultipleSpans() throws ExecutionException, InterruptedException {
         TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
         ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
-        AtomicReference<Span> currentSpanRefThread1 = new AtomicReference<>();
-        AtomicReference<Span> currentSpanRefThread2 = new AtomicReference<>();
-        AtomicReference<Span> currentSpanRefAfterEndThread2 = new AtomicReference<>();
-
-        AtomicReference<Span> parentSpanRef = new AtomicReference<>();
-        AtomicReference<Span> spanRef = new AtomicReference<>();
-        AtomicReference<Span> spanT2Ref = new AtomicReference<>();
 
         ThreadContextBasedTracerContextStorage spanTracerStorage = new ThreadContextBasedTracerContextStorage(
             threadContext,
@@ -366,52 +336,38 @@ public class DefaultTracerTests extends OpenSearchTestCase {
         );
         DefaultTracer defaultTracer = new DefaultTracer(tracingTelemetry, spanTracerStorage);
 
-        Future<?> asyncTask = executorService.submit(() -> {
+        CompletableFuture<?> asyncTask = CompletableFuture.runAsync(() -> {
             // create a parent span
             Span parentSpan = defaultTracer.startSpan(new SpanCreationContext("p_span_name", Attributes.EMPTY));
             SpanScope parentSpanScope = defaultTracer.withSpanInScope(parentSpan);
-            parentSpanRef.set(parentSpan);
             // create a span
             Span span = defaultTracer.startSpan(new SpanCreationContext("span_name_t_1", Attributes.EMPTY));
             SpanScope spanScope = defaultTracer.withSpanInScope(span);
-            spanRef.set(span);
 
-            Future<?> asyncTask1 = executorService.submit(() -> {
+            CompletableFuture<?> asyncTask1 = CompletableFuture.runAsync(() -> {
                 Span spanT2 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
                 SpanScope spanScopeT2 = defaultTracer.withSpanInScope(spanT2);
                 Span spanT21 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
                 SpanScope spanScopeT21 = defaultTracer.withSpanInScope(spanT21);
-                spanT2Ref.set(spanT21);
-                currentSpanRefThread2.set(defaultTracer.getCurrentSpan().getSpan());
+                assertEquals(spanT21, defaultTracer.getCurrentSpan().getSpan());
                 spanScopeT21.close();
                 spanT21.endSpan();
 
                 spanScopeT2.close();
                 spanT2.endSpan();
 
-                currentSpanRefAfterEndThread2.set(getCurrentSpanFromContext(defaultTracer));
-            });
-            try {
-                asyncTask1.get();
-            } catch (InterruptedException e) {
-                throw new RuntimeException(e);
-            } catch (ExecutionException e) {
-                throw new RuntimeException(e);
-            }
+                assertEquals(null, defaultTracer.getCurrentSpan());
+            }, executorService);
+
+            asyncTask1.join();
+
             spanScope.close();
             span.endSpan();
             parentSpanScope.close();
             parentSpan.endSpan();
-            currentSpanRefThread1.set(getCurrentSpanFromContext(defaultTracer));
-        });
+            assertEquals(null, defaultTracer.getCurrentSpan());
+        }, executorService);
         asyncTask.get();
-        assertEquals(spanT2Ref.get(), currentSpanRefThread2.get());
-        assertEquals(null, currentSpanRefAfterEndThread2.get());
-        assertEquals(null, currentSpanRefThread1.get());
-    }
-
-    private static Span getCurrentSpanFromContext(DefaultTracer defaultTracer) {
-        return defaultTracer.getCurrentSpan() != null ? defaultTracer.getCurrentSpan().getSpan() : null;
     }
 
     public void testClose() throws IOException {

